### PR TITLE
feat: add tags settings

### PR DESCRIPTION
**Pull request by zalevskyi**
[Task description](https://trello.com/c/KvBARoVc/21-add-an-ability-to-manage-tags-in-settings)

### CODE REVIEW

File: backend/src/data/entities/tag.ts [hunk 1]
```diff
@@ -1,26 +1,20 @@
...
 import { Page } from './page';
 import { Workspace } from './workspace';
 
 @Entity()
 export class Tag extends AbstractEntity {
   @RelationId((tag: Tag) => tag.workspace)
-  @Column()
-  readonly workspaceId: string;
+  @Column({ update: false })
+  workspaceId: string;
```
Comment (AlikRakhmonov) [inline backend/src/data/entities/tag.ts]: rollback, it was correct

File: backend/src/data/entities/tag.ts [hunk 2]
```diff
@@ -1,26 +1,20 @@
...
 
-  @ManyToOne(() => Workspace, workspace => workspace.tags)
+  @ManyToOne(() => Workspace, (workspace) => workspace.tags)
```
Comment (AlikRakhmonov) [inline backend/src/data/entities/tag.ts]: rollback

File: backend/src/data/entities/tag.ts [hunk 3]
```diff
@@ -1,26 +1,20 @@
...
   workspace: Workspace;
 
   @Column()
   name: string;
 
-  @ManyToMany(() => Page, page => page.tags)
+  @ManyToMany(() => Page, (page) => page.tags)
```
Comment (AlikRakhmonov) [inline backend/src/data/entities/tag.ts]: rollback

File: backend/src/services/tag.service.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  workspaceId: string,
+): Promise<ITag[]> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  const tags = await tagRepository.findAllByWorkspaceId(workspaceId);
+  return tags.map(({ id, workspaceId, name }) => ({ id, workspaceId, name }));
+};
+
+export const create = async (newTag: ITagCreation): Promise<ITag> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  if (newTag.name.length === 0) {
```
Comment (AlikRakhmonov) [inline backend/src/services/tag.service.ts]:
```suggestion
  if (!newTag.name) {
```

**Review by AlikRakhmonov — CHANGES_REQUESTED** [review]

File: frontend/src/store/tags/slice.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  isOpenNewTagForm: false,
+};
+
+const { reducer, actions } = createSlice({
+  name: ReducerName.TAGS,
+  initialState,
+  reducers: {
+    [ActionType.SetTags]: (state, action: PayloadAction<ITag[]>) => {
+      state.tags = action.payload;
+      // state.tagToEdit = null;
```
Comment (aleksandr-kuharenko) [inline frontend/src/store/tags/slice.ts]: remove unused code

**Review by aleksandr-kuharenko — APPROVED** [review]

File: backend/src/data/repositories/tag.repository.ts [hunk 6]
```diff
@@ -6,6 +6,9 @@ class TagRepository extends Repository<Tag> {
   public findById(id: string): Promise<Tag> {
     return this.findOne({ id });
   }
+  public findAllByWorkspaceId(workspaceId: string): Promise<Tag[]> {
+    return this.find({ workspaceId });
+  }
```
Comment (ursarik) [inline backend/src/data/repositories/tag.repository.ts]:
```suggestion
  public findById(id: string): Promise<Tag> {
    return this.findOne({ id });
  }
  
  public findAllByWorkspaceId(workspaceId: string): Promise<Tag[]> {
    return this.find({ workspaceId });
  }
```

File: backend/src/services/tag.service.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import { HttpError } from '../common/errors/http-error';
+import { HttpCode } from '../common/enums/http-code';
+import { HttpErrorMessage } from '../common/enums/http-error-message';
+
+export const getAllByWorkspaceId = async (
+  workspaceId: string,
+): Promise<ITag[]> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  const tags = await tagRepository.findAllByWorkspaceId(workspaceId);
+  return tags.map(({ id, workspaceId, name }) => ({ id, workspaceId, name }));
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]: shouldn't it be in separate file? in `backend/src/common/mappers`

File: backend/src/services/tag.service.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+export const create = async (newTag: ITagCreation): Promise<ITag> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  if (!newTag.name) {
+    throw new HttpError({
+      status: HttpCode.BAD_REQUEST,
+      message: HttpErrorMessage.TAG_EMPTY_STRING,
+    });
+  }
+  const tagsInDB = await tagRepository.find(newTag);
+  if (tagsInDB.length > 0) {
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]:
```suggestion
  if (tagsInDB.length) {
```

File: backend/src/services/tag.service.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    });
+  }
+  const { id, workspaceId, name } = await tagRepository.save(newTag);
+  return { id, workspaceId, name };
+};
+
+export const deleteById = async (id: string): Promise<void> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  const result = await tagRepository.softDelete(id);
+  if (result.affected === 0) {
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]:
```suggestion
  if (!result.affected) {
```

File: backend/src/services/tag.service.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    throw new HttpError({
+      status: HttpCode.NOT_FOUND,
+      message: HttpErrorMessage.TAG_IN_WORKSPACE_NOT_FOUND,
+    });
+  }
+  return;
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]:
```suggestion
```

File: backend/src/services/tag.service.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+): Promise<{ id: string; name: string }> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  if (!name) {
+    throw new HttpError({
+      status: HttpCode.BAD_REQUEST,
+      message: HttpErrorMessage.TAG_EMPTY_STRING,
+    });
+  }
+  const result = await tagRepository.update(id, { name });
+  if (result.affected === 0) {
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]:
```suggestion
  if (!result.affected) {
```

File: backend/src/services/tag.service.ts (hunk 7 above)
Comment (zalevskyi) [reply backend/src/services/tag.service.ts]: Not exactly.
I prefer to use built in RDBMS query tools, rather then inventing mapping wheels
Redone repository

File: backend/src/services/tag.service.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+import TagRepository from '../data/repositories/tag.repository';
+import { HttpError } from '../common/errors/http-error';
+import { HttpCode } from '../common/enums/http-code';
+import { HttpErrorMessage } from '../common/enums/http-error-message';
+
+export const getAllByWorkspaceId = async (
+  workspaceId: string,
+): Promise<ITag[]> => {
+  const tagRepository = getCustomRepository(TagRepository);
+  return await tagRepository.findAllByWorkspaceId(workspaceId);
```
Comment (ursarik) [inline backend/src/services/tag.service.ts]:
```suggestion
  return tagRepository.findAllByWorkspaceId(workspaceId);
```

**Review by AlikRakhmonov — APPROVED** [review]

**Review by DeniGolf — APPROVED** [review]

**Pull request merged**